ding/policy/mbpolicy/mbsac.py

```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
+            pred = dist.rsample()
+            action = torch.tanh(pred)
+
+            log_prob = dist.log_prob(
+                pred
+            ) + 2 * (pred + torch.nn.functional.softplus(-2. * pred) - torch.log(torch.tensor(2.))).sum(-1)
+            aug_reward = -self._alpha.detach() * log_prob
+
+            return unfold_batch(action, dim), unfold_batch(aug_reward, dim)
+
+        self._actor_fn = actor_fn
+
+        def critic_fn(obss: Tensor, actions: Tensor, model: nn.Module):
+            eval_data = {'obs': obss, 'action': actions}
+            q_values = model.forward(eval_data, mode='compute_critic')['q_value']
+            return q_values
+
+        self._critic_fn = critic_fn
+
+    def _forward_learn(self, data: dict, world_model, envstep) -> Dict[str, Any]:
+        # preprocess data
+        data = default_preprocess_learn(
+            data,
+            use_priority=self._priority,
+            use_priority_IS_weight=self._cfg.priority_IS_weight,
+            ignore_done=self._cfg.learn.ignore_done,
+            use_nstep=False
+        )
+        if self._cuda:
+            data = to_device(data, self._device)
+
+        if len(data['action'].shape) == 1:
+            data['action'] = data['action'].unsqueeze(1)
+
+        # [B, D] -> [E, B, D]
+        data['next_obs'] = data['next_obs'].unsqueeze(0).repeat(self._ensemble_size, 1, 1)
+        data['reward'] = data['reward'].unsqueeze(0).repeat(self._ensemble_size, 1)
+        data['done'] = data['done'].unsqueeze(0).repeat(self._ensemble_size, 1)
+
+        self._learn_model.train()
+        self._target_model.train()
+
+        obss, actions, rewards, aug_rewards, dones = \
+            world_model.rollout(data['next_obs'], self._actor_fn, envstep, keep_ensemble=True)
+        rewards = torch.concat([data['reward'].unsqueeze(0), rewards])
+        dones = torch.concat([data['done'].unsqueeze(0), dones])
+
+        # (T, E, B)
+        target_q_values = q_evaluation(obss, actions, partial(self._critic_fn, model=self._target_model))
+        if self._twin_critic:
+            target_q_values = torch.min(target_q_values[0], target_q_values[1]) + aug_rewards
+        else:
+            target_q_values = target_q_values + aug_rewards
+
+        # (T+1, E, B)
+        discounts = ((1 - dones) * self._gamma).cumprod(dim=0)
+        discounts = torch.concat([torch.ones_like(discounts)[:1], discounts])
+        # (T, E, B)
+        cum_rewards = (rewards * discounts[:-1]).cumsum(dim=0)
+        discounted_q_values = target_q_values * discounts[1:]
+        steve_return = cum_rewards + discounted_q_values
+        # (T, B)
+        steve_return_mean = steve_return.mean(1)
+        with torch.no_grad():
+            steve_return_inv_var = 1 / (1e-8 + steve_return.var(1, unbiased=False))
+            steve_return_weight = steve_return_inv_var / (1e-8 + steve_return_inv_var.sum(dim=0))
+        # (B)
+        steve_return = (steve_return_mean * steve_return_weight).sum(0)
```

Comment:
I tested the program logic of calculating STEVE target return. Well done! 
Compared with the origin paper, we didn't have an additional ensemble of reward model and followed MBPO ensemble form. So the ensemble calculation is clear.

Reply:
We do have an ensemble of reward models but they are actually coupled with the dynamics model. 
We do not have a Q ensemble and we also do not plan to have a Q ensemble.